File: src/http-client-mock/MockHandler.cs
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;

using Matcher = System.Func<System.Net.Http.HttpRequestMessage, bool>;
using Responder = System.Func<System.Net.Http.HttpRequestMessage, System.Net.Http.HttpResponseMessage>;

namespace HttpClientMock
{
    internal class MockHandler : DelegatingHandler
    {
        private readonly (Matcher Matcher, Responder Responder)[] behaviors;

        public MockHandler((Matcher Matcher, Responder Responder)[] behaviors)
        {
            this.behaviors = behaviors;
        }

        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, System.Threading.CancellationToken cancellationToken)
        {
            var httpResponseMessage = CreateResponse(request);
            if (request.Method == HttpMethod.Head)
            {
                httpResponseMessage.Content = null;
            }
            return Task.FromResult(httpResponseMessage);
        }
        
        private HttpResponseMessage CreateResponse(HttpRequestMessage request)
        {
            var (_, responder) = behaviors.Reverse().FirstOrDefault(
                behavior => behavior.Matcher(request));
            
            return responder != null ? responder(request) 
                : new HttpResponseMessage(HttpStatusCode.NotFound);
        }

    }
}

File: src/http-client-mock/MockHandler.fs
namespace HttpClientMock

open System.Net
open System.Net.Http
open System.Threading.Tasks


type Behavior = {Matcher: HttpRequestMessage -> bool; Responder: HttpRequestMessage -> HttpResponseMessage}

type MockHandler(behaviors: Behavior[]) =
    inherit DelegatingHandler()
    
    override this.SendAsync(request: HttpRequestMessage, _: System.Threading.CancellationToken) =
        match behaviors |> Seq.tryFindBack(fun b -> b.Matcher(request)) with
                | None -> new HttpResponseMessage(HttpStatusCode.NotFound)
                | Some(b) -> b.Responder(request)
            |> fun r -> if request.Method = HttpMethod.Head then r.Content <- null
                        r
            |> Task.FromResult
